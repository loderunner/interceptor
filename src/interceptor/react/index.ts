export { Provider as InterceptProvider, useIntercepts } from './intercept'
