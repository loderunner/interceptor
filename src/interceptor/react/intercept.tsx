import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
