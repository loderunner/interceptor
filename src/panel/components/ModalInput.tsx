import * as React from 'react'
import { useCallback, useEffect, useMemo, useRef, useState } from 'react'

import type { CSSProperties, ChangeEvent, KeyboardEvent } from 'react'

interface Props {
  element: HTMLElement
  value: string
  onChange: (value: string) => void
  onCancel: () => void
}

const ModalInput = ({ element, value, onChange, onCancel }: Props) => {
  const [inputValue, setInputValue] = useState(value)
  const ref = useRef<HTMLInputElement>(null)

  useEffect(() => {
    ref.current?.focus()
    ref.current?.select()
  }, [])

  const onChangeInput = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => setInputValue(e.target.value),
    []
  )

  const onBlur = useCallback(() => onChange(inputValue), [inputValue])

  const onKeyDown = useCallback(
    (e: KeyboardEvent<HTMLInputElement>) => {
      switch (e.key) {
        case 'Enter':
          onChange(inputValue)
          break
        case 'Escape':
          onCancel()
          break
        default:
          return
      }
      e.stopPropagation()
    },
    [inputValue]
  )

  // Inline styles from element properties
  const inlineStyles = useMemo((): CSSProperties => {
    const rect = element.getBoundingClientRect()
    const style = getComputedStyle(element)
    return {
      fontSize: style.fontSize,
      lineHeight: style.lineHeight,
      marginLeft: `calc(-0.25rem - ${style.paddingLeft})`,
      marginTop: `calc(-0.25rem - ${style.paddingTop})`,
      padding: '0.25rem',
      left: `${rect.left}px`,
      top: `${rect.top}px`,
      width: `calc(${rect.width}px + 2 * 0.25rem)`,
    }
  }, [element])

  return (
    <input
      className="fixed"
      style={inlineStyles}
      type="text"
      value={inputValue}
      onChange={onChangeInput}
      onBlur={onBlur}
      onKeyDown={onKeyDown}
      ref={ref}
    />
  )
}

export default ModalInput
