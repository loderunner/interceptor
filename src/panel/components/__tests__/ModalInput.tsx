import { cleanup, fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'
import { useEffect, useRef, useState } from 'react'
import '@testing-library/jest-dom'

import ModalInput from '../ModalInput'

describe('[ModalInput]', () => {
  const onChange = jest.fn()
  const onCancel = jest.fn()
  beforeEach(() => {
    jest.resetAllMocks()
  })

  afterEach(() => {
    cleanup()
  })

  it('should match empty snapshot', async () => {
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { container, findByRole } = render(
      <ModalInput
        element={spanContainer}
        value=""
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    await findByRole('textbox')

    expect(container).toMatchSnapshot()
  })

  it('should have focus and be selected', async () => {
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    const input = (await findByRole('textbox')) as HTMLInputElement

    expect(input).toHaveValue('toto')
    expect(input).toHaveFocus()
    expect(input.selectionStart).toBe(0)
    expect(input.selectionEnd).toBe(4)
  })

  it('should change value on change event', async () => {
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    const input = (await findByRole('textbox')) as HTMLInputElement

    fireEvent.change(input, { target: { value: 'hello world' } })

    expect(input).toHaveValue('hello world')
  })

  it('should change value on keyboard events', async () => {
    const user = userEvent.setup()
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    const input = (await findByRole('textbox')) as HTMLInputElement

    await user.keyboard('hello world')

    expect(input).toHaveValue('hello world')
  })

  it('should fire onChange event on blur', async () => {
    const user = userEvent.setup()
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    await findByRole('textbox')

    await user.keyboard('hello world')
    await user.click(spanContainer as HTMLSpanElement)

    expect(onChange).toBeCalledWith('hello world')
    expect(onCancel).not.toBeCalled()
  })

  it('should fire onChange event on Enter key down', async () => {
    const user = userEvent.setup()
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    await findByRole('textbox')

    await user.keyboard('hello world')
    await user.keyboard('{Enter}')

    expect(onChange).toBeCalledWith('hello world')
    expect(onCancel).not.toBeCalled()
  })

  it('should fire onCancel event on Esc key down', async () => {
    const user = userEvent.setup()
    const { container: spanContainer } = render(<span>Hello World!</span>)
    const { findByRole } = render(
      <ModalInput
        element={spanContainer}
        value="toto"
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    await findByRole('textbox')

    await user.keyboard('hello world')
    await user.keyboard('{Esc}')

    expect(onChange).not.toBeCalled()
    expect(onCancel).toBeCalled()
  })
})
