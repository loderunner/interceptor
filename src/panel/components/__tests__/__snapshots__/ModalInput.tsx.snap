// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ModalInput] should match empty snapshot 1`] = `
<div>
  <input
    class="fixed "
    style="padding: 0.25rem; left: 0px; top: 0px; width: calc(0px + 2 * 0.25rem); min-width: 120px;"
    type="text"
    value=""
  />
</div>
`;
